### PR TITLE
Make alpha test overrides use the same system as blend mode overrides

### PR DESCRIPTION
**Pull request by maximumpower55**
This will close this: #219
Don't merge yet big changes

### CODE REVIEW

File: src/main/java/net/coderbot/iris/gl/blending/AlphaTest.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package net.coderbot.iris.gl.blending;
+
+public class AlphaTest {
+	private final int func;
```
[the remark this answers is not in the document]
Comment (coderbot16) [reply src/main/java/net/coderbot/iris/gl/blending/AlphaTest.java:4]: Yes

File: src/main/java/net/coderbot/iris/gl/blending/AlphaTest.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	private final float reference;
+
+	public AlphaTest(int func, float reference) {
```
[the remark this answers is not in the document]
Comment (coderbot16) [reply src/main/java/net/coderbot/iris/gl/blending/AlphaTest.java:7]: Yes

File: src/main/java/net/coderbot/iris/gl/blending/AlphaTestOverride.java [hunk 3]
```diff
@@ -1,44 +1,19 @@
...
-		public void setup() {
-			GlStateManager._disableAlphaTest();
-		}
-
-		@Override
-		public String toString() {
-			return "AlphaTestOverride { off }";
-		}
+	public static void restore() {
+		AlphaTestStorage.restoreBlend();
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/blending/AlphaTestOverride.java:17]:
```suggestion
		AlphaTestStorage.restoreAlpha();
```

File: src/main/java/net/coderbot/iris/gl/blending/AlphaTestStorage.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+	public static void deferAlphaTestToggle(boolean enabled) {
+		originalAlphaTestEnable = enabled;
+	}
+
+	public static void deferAlphaFunc(int func, float reference) {
+		originalAlphaTest = new AlphaTest(func, reference);
+	}
+
+	public static void restoreBlend() {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/blending/AlphaTestStorage.java:45]:
```suggestion
	public static void restoreAlphaTest() {
```

File: src/main/java/net/coderbot/iris/mixin/GlStateManagerAccessor.java [hunk 5]
```diff
@@ -6,6 +6,11 @@
 
 @Mixin(GlStateManager.class)
 public interface GlStateManagerAccessor {
+	@Accessor("ALPHA_TEST")
+	static GlStateManager.AlphaState getALPHA_TEST() {
```
[the remark this answers is not in the document]
Comment (coderbot16) [reply src/main/java/net/coderbot/iris/mixin/GlStateManagerAccessor.java:10]: This is probably fine as long as it's consistent with other accessors we do

File: src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(GlStateManager.class)
+public class MixinGlStateManager_AlphaTestOverride {
+	@Inject(method = "_disableAlphaTest", at = @At("HEAD"), cancellable = true)
+	private static void iris$alphaTestDisableLock(CallbackInfo ci) {
+		if(AlphaTestStorage.isAlphaTestLocked()) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java:15]:
```suggestion
		if (AlphaTestStorage.isAlphaTestLocked()) {
```

File: src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+			AlphaTestStorage.deferAlphaTestToggle(true);
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java:16]: You might have swapped this

File: src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+			ci.cancel();
+		}
+	}
+
+	@Inject(method = "_enableAlphaTest", at = @At("HEAD"), cancellable = true)
+	private static void iris$alphaTestEnableLock(CallbackInfo ci) {
+		if (AlphaTestStorage.isAlphaTestLocked()) {
+			AlphaTestStorage.deferAlphaTestToggle(false);
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinGlStateManager_AlphaTestOverride.java:24]: You might have swapped this

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Thanks, please see comments.

**Comment by maximumpower55** [comment]
I have made most of the requested changes locally but the type change of function has really stumped me

**Comment by maximumpower55** [comment]
I can't figure out how to convert a gl id into the alpha test function enum

**Comment by coderbot16** [comment]
It might be necessary to add a conversion function from GL id to enum (maybe using a switch / Map) to AlphaTestFunction or whatever the class is.

**Comment by maximumpower55** [comment]
How did i not think of that well going to implement

**Comment by maximumpower55** [comment]
Made requested changes

**Review by coderbot16 — APPROVED** [review]
Thanks!

**Pull request merged**